pkg/resource/deploy/step_generator.go

```diff
@@ -44,6 +44,8 @@ type stepGenerator struct {
 	deployment *Deployment // the deployment to which this step generator belongs
 	opts       Options     // options for this step generator
 
+	urnLock *resourceLock // Used for URN based locking
```

Comment:
looks like this is doing double-duty: it's both a global lock for the state of the step generator and a set of granular locks for each URN. is that correct?

Reply:
Yes, they have to be done together to prevent lock inversion